campaignion_tracking/campaignion_tracking.module

```diff
@@ -112,12 +118,14 @@ function campaignion_tracking_webform_redirect_alter($redirect, $submission) {
     $submissions[$submission->sid] = $submission->unwrap();
     webform_paymethod_select_webform_submission_load($submissions);
 
-    $fragment = 't=ds';
+    // Track a donation success (ds).
+    $fragments[] = 't=ds';
     foreach ($submissions[$submission->sid]->payments as $payment) {
-      // Prefix `dm` for donation method.
-      $fragment .= ";d:m=" . $payment->method->name;
+      // Add context for donation tracking.
+      $fragments[] = "d:m=" . $payment->method->name;
     }
   }
 
-  $redirect->fragment = empty($redirect->fragment) ? $fragment : $redirect->fragment . ';' . $fragment;
+  $tracking_fragment = implode(';', $fragments);
+  $redirect->fragment = empty($redirect->fragment) ? $tracking_fragment : $redirect->fragment . ';' . $tracking_fragment;
```

Comment:
How does this work wrt url encoding? Why are the values individually prefixed instead of using `&`-separated key value pairs?

Reply:
Is there a convention to use `&` in our fragment encoding?

Reply:
Can you give an example how you would suggest to encode the information in the fragment?

Reply:
The idea was to have something like this:

```
#real-fragment-without-namespace;namespace1:key1=value1&key2=value2;namespace2:key1=value1&…
```

Reply:
I see.

Although there is not much difference besides brevity to
```
#real-fragment-without-namespace;namespace1:key1=value1;namespace1:key2=value2;namespace2:key1=value1;…
```

- Without `&` parsing/encoding is simpler.
- Namespaces in this module are quite short (`t`, `d`, `w`), so not much added length
- Concepts stay the same

Reply:
We probably can just add splitting by `&` to the parser. If namespaces appear multiple times, they just get merged. This is currently how the tracking code work. Easier to just add additional information then...

Reply:
This scheme was invented with webform_prefill. I’d like to keep it consistent across projects.

Reply:
I don’t think parsing is simpler: This way parsing is.

1. split by `;`
2. Find parts that start with the right `prefix:`.
3. Use a parser for URL-parameters (usually in the standard library) to parse the string and get all key value pairs and decode them.